app/com/arpnetworking/metrics/portal/scheduling/TwoWayJobRefSerializer.java

```diff
@@ -0,0 +1,122 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.arpnetworking.metrics.portal.scheduling;
+
+import com.arpnetworking.steno.Logger;
+import com.arpnetworking.steno.LoggerFactory;
+import com.google.common.collect.ImmutableMap;
+import models.internal.impl.DefaultOrganization;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.function.Function;
+
+/**
+ * A {@link JobRefSerializer} that allows construction to and from entity IDs.
+ *
+ * Round-trip is only guaranteed for refs whose repository types are whitelisted by this serializer
+ * (e.g. allowed by the constructor arguments).
+ *
+ * @author Christian Briones (cbriones at dropbox dot com)
+ */
+public class TwoWayJobRefSerializer implements JobRefSerializer {
+    private static final Logger LOGGER = LoggerFactory.getLogger(TwoWayJobRefSerializer.class);
+
+    private final Map<String, Class<? extends JobRepository<?>>> _repositories;
+    private final Map<String, Class<? extends JobExecutionRepository<?>>> _execRepositories;
+
+    /**
+     * Public constructor.
+     *
+     * @param allowedJobRepositories the list of allowed job repositories
+     * @param allowedJobExecRepositories the list of allowed job execution repositories.
+     */
+    public TwoWayJobRefSerializer(
+            final List<Class<? extends JobRepository<?>>> allowedJobRepositories,
+            final List<Class<? extends JobExecutionRepository<?>>> allowedJobExecRepositories
+    ) {
+        _repositories = allowedJobRepositories.stream()
+                .collect(ImmutableMap.toImmutableMap(
+                        Class::getSimpleName,
+                        Function.identity()
+                ));
+        _execRepositories = allowedJobExecRepositories.stream()
+                .collect(ImmutableMap.toImmutableMap(
+                        Class::getSimpleName,
+                        Function.identity()
+                ));
+    }
+
+    @Override
+    public String jobRefToEntityID(final JobRef<?> ref) {
+        final String repoName = ref.getRepositoryType().getSimpleName();
+        if (!_repositories.containsKey(repoName)) {
+            LOGGER.warn()
+                .setMessage("The repository on this ref is not whitelisted for serialization.")
+                .addData("repositoryType", ref.getRepositoryType());
+        }
+        final String execRepoName = ref.getExecutionRepositoryType().getSimpleName();
+        if (!_execRepositories.containsKey(execRepoName)) {
+            LOGGER.warn()
+                    .setMessage("The execution repository on this ref is not whitelisted for serialization.")
+                    .addData("repositoryType", ref.getExecutionRepositoryType());
+        }
+        return String.join(
+                "_",
```

Comment:
Technically I believe underscore is allowed in class and package names in Java.

https://www.geeksforgeeks.org/character-isjavaidentifierpart-method-in-java-with-examples/
https://docs.oracle.com/javase/7/docs/api/java/lang/Character.html